<a href="https://colab.research.google.com/github/emely3h/Geospatial_ML/blob/main/experiments/experiment_8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# U-Net Experiment 8: Finding optimal depth of U-Net

### 0. Prepare Colab, Define Constants

In [1]:
from google.colab import drive

drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
#! ls
%cd drive/MyDrive/MachineLearning/
#! git clone https://github.com/emely3h/Geospatial_ML.git
%cd Geospatial_ML
! ls

/content/drive/MyDrive/MachineLearning
/content/drive/MyDrive/MachineLearning/Geospatial_ML
data_exploration  experiments	     models	   pyproject.toml    scripts
docs		  image_processing   poetry.lock   README.md
evaluation	  metrics_bug.ipynb  prepare_data  requirements.txt


In [3]:
import numpy as np
import os
import pandas as pd
import tensorflow as tf
from keras.losses import categorical_crossentropy
from tensorflow.keras.callbacks import EarlyStopping
import pickle
from keras.utils import Sequence
from datetime import datetime
from data_exploration.mask_stats import Mask_Stats
from tensorflow.keras.callbacks import ModelCheckpoint
from tensorflow.keras.optimizers.experimental import Adamax

In [4]:
total_tiles = 11121
train_tiles = 6672
test_val_tiles = 2224
data_path = "../data_colab/256_256"
experiment = "experiment_8"
batch_size = 32
tile_size = 256
step_size = 256

### 1. Create Data Generators

In [ ]:
train_split_x = np.memmap(os.path.join(data_path, "train_split_x.npy"), mode="r", shape=(train_tiles, 256, 256, 5),
                          dtype=np.uint8)
train_split_y = np.memmap(os.path.join(data_path, "train_split_y.npy"), mode="r", shape=(train_tiles, 256, 256),
                          dtype=np.uint8)
val_split_x = np.memmap(os.path.join(data_path, "val_split_x.npy"), mode="r", shape=(test_val_tiles, 256, 256, 5),
                        dtype=np.uint8)
val_split_y = np.memmap(os.path.join(data_path, "val_split_y.npy"), mode="r", shape=(test_val_tiles, 256, 256),
                        dtype=np.uint8)
test_split_x = np.memmap(os.path.join(data_path, "test_split_x.npy"), mode="r", shape=(test_val_tiles, 256, 256, 5),
                         dtype=np.uint8)
test_split_y = np.memmap(os.path.join(data_path, "test_split_y.npy"), mode="r", shape=(test_val_tiles, 256, 256),
                         dtype=np.uint8)

train_stats = Mask_Stats(train_split_y)
train_stats.print_stats()
print()
val_stats = Mask_Stats(val_split_y)
val_stats.print_stats()
print()
test_stats = Mask_Stats(test_split_y)
test_stats.print_stats()

Shape: (6672, 256, 256)
Land pixels: 195058814  44.610 %
Valid pixels: 138904480  31.767 %
Invalid pixels: 103292898  23.623 %
Sum: 6672

Shape: (2224, 256, 256)
Land pixels: 65320265  44.816 %
Valid pixels: 46246663  31.730 %
Invalid pixels: 34185136  23.454 %
Sum: 2224

Shape: (2224, 256, 256)
Land pixels: 64786699  44.450 %
Valid pixels: 46892391  32.173 %
Invalid pixels: 34072974  23.377 %
Sum: 2224


In [ ]:
class DataGenerator(Sequence):
    def __init__(self, mmap_x, mmap_y, batch_size):
        self.x_input = mmap_x
        self.y_mask = mmap_y
        self.batch_size = batch_size
        self.num_samples = self.x_input.shape[0]

    # returns number of batches as int
    def __len__(self):
        return int(np.ceil(self.num_samples / float(self.batch_size)))

    # returns single batch
    def __getitem__(self, index):
        batch_indices = slice(index * self.batch_size, (index + 1) * self.batch_size)
        batch_inputs = self.x_input[batch_indices]
        batch_masks = self.y_mask[batch_indices]

        # normalization
        batch_inputs = batch_inputs / 255
        # one-hot-encoding
        batch_masks = np.array([tf.one_hot(item, depth=3).numpy() for item in batch_masks])

        # normalization + one hot encoding
        return batch_inputs, batch_masks

    def getitem_as_img(self, index):
        batch_indices = slice(index * self.batch_size, (index + 1) * self.batch_size)
        batch_inputs = self.x_input[batch_indices]
        batch_masks = self.y_mask[batch_indices]
        # normalization + one hot encoding
        return batch_inputs, batch_masks

In [ ]:
# instanciate DataGenerators
batch_size = 32

train_generator = DataGenerator(train_split_x, train_split_y, batch_size)
val_generator = DataGenerator(val_split_x, val_split_y, batch_size)
test_generator = DataGenerator(test_split_x, test_split_y, batch_size)

print(train_generator.__len__())
print(val_generator.__len__())
print(test_generator.__len__())

209
70
70


In [ ]:
train_batch = train_generator.__getitem__(9)
val_batch = val_generator.__getitem__(3)
test_batch = test_generator.__getitem__(4)


def print_batch_shapes(batch):
    print(type(batch))
    print(batch[0].shape)
    print(batch[1].shape)
    print()


# check batch shapes
print_batch_shapes(train_batch)
print_batch_shapes(val_batch)
print_batch_shapes(test_batch)

# check normalization
print('Check normalization')
print(train_batch[1].max())
print(train_batch[1].min())

print(val_batch[1].max())
print(val_batch[1].min())

print(test_batch[1].max())
print(test_batch[1].min())

print()
# check one-hot-encoding
print('check one hot encoding')
print(train_batch[0].max())
print(train_batch[0].min())

print(val_batch[0].max())
print(val_batch[0].min())

print(test_batch[0].max())
print(test_batch[0].min())

<class 'tuple'>
(32, 256, 256, 5)
(32, 256, 256, 3)

<class 'tuple'>
(32, 256, 256, 5)
(32, 256, 256, 3)

<class 'tuple'>
(32, 256, 256, 5)
(32, 256, 256, 3)

Check normalization
1.0
0.0
1.0
0.0
1.0
0.0

check one hot encoding
1.0
0.0
1.0
0.0
1.0
0.0


### 3. Model training
execute with premium GPU, High RAM

In [ ]:
!nvidia-smi

Tue May 16 20:58:50 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.85.12    Driver Version: 525.85.12    CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A100-SXM...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   31C    P0    48W / 400W |    693MiB / 40960MiB |      2%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

In [ ]:
from tensorflow import keras
from tensorflow.keras import layers


def build_unet_model(input_shape, num_layers, num_classes=3):
    inputs = keras.Input(shape=input_shape)
    x = inputs

    # Encoder
    skips = []
    filters = 16
    for _ in range(num_layers):
        x = layers.Conv2D(filters, 3, activation='relu', padding='same')(x)
        x = layers.Conv2D(filters, 3, activation='relu', padding='same')(x)
        skips.append(x)
        x = layers.MaxPooling2D(2)(x)
        filters *= 2

    # Bridge
    x = layers.Conv2D(filters, 3, activation='relu', padding='same')(x)
    x = layers.Conv2D(filters, 3, activation='relu', padding='same')(x)

    # Decoder
    skips = reversed(skips)
    filters //= 2
    for skip in skips:
        x = layers.Conv2DTranspose(filters, 2, strides=2, activation='relu', padding='same')(x)
        x = layers.Concatenate()([skip, x])
        x = layers.Conv2D(filters, 3, activation='relu', padding='same')(x)
        x = layers.Conv2D(filters, 3, activation='relu', padding='same')(x)
        filters //= 2

    # Output
    outputs = layers.Conv2D(num_classes, (1, 1), activation='softmax')(x)

    model = keras.Model(inputs, outputs)
    return model

In [ ]:
all_metrics = []

for num_layers in range(1, 8):
    count = num_layers + 8
    model_metrics = []
    model_name = f'{tile_size}_{step_size}_run_{count}'
    model_path = f'../models/{experiment}/model_{model_name}.h5'

    print(f'{count} num_layers: {num_layers} Started at: {datetime.now()}')

    # Define the mIoU metric
    mean_iou = tf.keras.metrics.OneHotMeanIoU(num_classes=3, name='mean_iou')
    invalid_iou = tf.keras.metrics.OneHotIoU(num_classes=3, target_class_ids=[0], name='invalid_iou')
    valid_iou = tf.keras.metrics.OneHotIoU(num_classes=3, target_class_ids=[1], name='valid_iou')
    land_iou = tf.keras.metrics.OneHotIoU(num_classes=3, target_class_ids=[2], name='land_iou')

    # compiling model
    model = build_unet_model((256, 256, 5), num_layers)
    model.compile(optimizer=Adamax(), loss=categorical_crossentropy,
                  metrics=[mean_iou, invalid_iou, valid_iou, land_iou, 'accuracy'])
    print(model.summary())

    # callbacks
    early_stop_loss = EarlyStopping(monitor='val_loss', mode='min', patience=15)
    early_stop_acc = EarlyStopping(monitor='val_mean_iou', mode='max', patience=15)
    checkpoint = ModelCheckpoint(model_path, monitor="val_mean_iou", mode="max", save_best_only=True, verbose=1)

    # training
    model_history = model.fit(x=train_generator, epochs=100, validation_data=val_generator,
                              callbacks=[early_stop_loss, early_stop_acc, checkpoint])

    # Save model history
    with open(f'../models/{experiment}/history_{model_name}.pkl', 'wb') as file_pi:
        pickle.dump(model_history.history, file_pi)
    print('saving history completed')

    print(f'{count} Finished at: {datetime.now()}')
    print(f'{count} Metrics')
    print(model.metrics_names)
    print('training metrics')
    train_metrics = model.evaluate(train_generator, verbose=2)
    print(train_metrics)
    print('validation metrics')
    val_metrics = model.evaluate(val_generator, verbose=2)
    print(val_metrics)
    print('test metrics')
    test_metrics = model.evaluate(test_generator, verbose=2)
    print(test_metrics)
    all_metrics.append(train_metrics)
    all_metrics.append(val_metrics)
    all_metrics.append(test_metrics)

    with open(f'../metrics/{experiment}_{count}.pkl', 'wb') as file_pi:
        pickle.dump(all_metrics, file_pi)
    print('saving metrics completed')

9 num_layers: 1 Started at: 2023-05-16 20:58:50.430259
Model: "model"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 input_1 (InputLayer)           [(None, 256, 256, 5  0           []                               
                                )]                                                                
                                                                                                  
 conv2d (Conv2D)                (None, 256, 256, 16  736         ['input_1[0][0]']                
                                )                                                                 
                                                                                                  
 conv2d_1 (Conv2D)              (None, 256, 256, 16  2320        ['conv2d[0][0]']                 
                                )      

KeyboardInterrupt: ignored

### 4. Metrics

In [8]:
metrics = []
files = os.listdir(f'../metrics/{experiment}/')

for i in range(len(files) - 1, -1, -1):
    if 'layers' in files[i]:
        del files[i]

files = sorted(files)

for metric_file in files:
    with open(f'../metrics/{experiment}/{metric_file}', 'rb') as file:
        metric_dict = pickle.load(file).__dict__
        metrics.append(metric_dict)

In [9]:
df = pd.DataFrame(metrics)

for i, _ in enumerate(files):
    files[i] = files[i][8:-4]

df.index = files
df

iou_invalid  iou_valid  iou_land  mean_iou  precision_invalid  \
test_1_1      0.855802   0.902306  0.996778  0.918295           0.959200   
test_2_1      0.511113   0.722353  0.736162  0.656543           0.528110   
test_2_3      0.877207   0.914738  0.997537  0.929828           0.956470   
test_3_1      0.898628   0.927567  0.998555  0.941583           0.953436   
test_3_2      0.899704   0.928168  0.998078  0.941984           0.950507   
test_3_3      0.900827   0.929737  0.998839  0.943134           0.962381   
test_4_1      0.908179   0.935095  0.998831  0.947368           0.967426   
test_4_2      0.909002   0.935425  0.998509  0.947645           0.961316   
test_4_3      0.910239   0.936735  0.998520  0.948498           0.965619   
test_5_1      0.904243   0.932276  0.997891  0.944803           0.957190   
test_5_2      0.898276   0.927773  0.998148  0.941399           0.953730   
test_6_1      0.886138   0.920380  0.997017  0.934511           0.954151   
test_6_2      0.880337   0.913813  0.997281  0.930477           0.935362   
test_7_1      0.876594   0.912148  0.996982  0.928575           0.942059   
test_7_2      0.878144   0.915548  0.996565  0.930086           0.952891   
test_8_1      0.872630   0.910255  0.997221  0.926702           0.945220   
train_1_1     0.835189   0.871764  0.996640  0.901198           0.946871   
train_2_1     0.857170   0.888351  0.997805  0.914442           0.957961   
train_2_3     0.865162   0.893439  0.997442  0.918681           0.955200   
train_3_1     0.907414   0.924174  0.998630  0.943406           0.957894   
train_3_2     0.908723   0.925250  0.997838  0.943937           0.956117   
train_3_3     0.908886   0.926300  0.998764  0.944650           0.967511   
train_4_1     0.930835   0.944167  0.998827  0.957943           0.977616   
train_4_2     0.935256   0.947366  0.998677  0.960433           0.973236   
train_4_3     0.934171   0.946838  0.998713  0.959907           0.977383   
train_5_1     0.934618   0.947363  0.997838  0.959940           0.972350   
train_5_2     0.933404   0.946477  0.998177  0.959353           0.972928   
train_6_1     0.911218   0.928983  0.996719  0.945640           0.964737   
train_6_2     0.934366   0.946734  0.997016  0.959372           0.960222   
train_7_1     0.882314   0.905077  0.996846  0.928079           0.950132   
train_7_2     0.882556   0.907856  0.996357  0.928923           0.961987   
train_8_1     0.939699   0.951841  0.997011  0.962850           0.970913   
val_1_1       0.843271   0.892588  0.996976  0.910945           0.960239   
val_2_1       0.861116   0.903738  0.997966  0.920940           0.964063   
val_2_3       0.866093   0.906259  0.997623  0.923325           0.959324   
val_3_1       0.893114   0.923248  0.998508  0.938290           0.958373   
val_3_2       0.894971   0.924408  0.997967  0.939115           0.956267   
val_3_3       0.894357   0.924702  0.998783  0.939281           0.966384   
val_4_1       0.902563   0.930743  0.998762  0.944023           0.971264   
val_4_2       0.905094   0.932326  0.998450  0.945290           0.967006   
val_4_3       0.903972   0.931920  0.998422  0.944771           0.970033   
val_5_1       0.902195   0.930227  0.997810  0.943410           0.961506   
val_5_2       0.894051   0.924189  0.998154  0.938798           0.958426   
val_6_1       0.881354   0.916058  0.997084  0.931499           0.957615   
val_6_2       0.877001   0.910883  0.997280  0.928388           0.941071   
val_7_1       0.872350   0.908795  0.996992  0.926046           0.948494   
val_7_2       0.872359   0.911111  0.996765  0.926745           0.960775   
val_8_1       0.865068   0.904913  0.997350  0.922443           0.954710   

           precision_valid  precision_land  mean_precision  recall_invalid  \
test_1_1          0.924799        0.998060        0.960687        0.888131   
test_2_1          0.947211        0.998782        0.824701        0.940762   
test_2_3          0.941093        0.998234        0.

Note that the indexes containing 'layers' in their name are additional

The table shows that the optimal number of layers is between 3 and 5. The following table therefore only forms the metrics for the test data for trainigs with 3-5 layers, whereby the average value for 3,4 and 5 layers is also calculated.

In [7]:
df_mean = df.iloc[3:11]
mean_3_layers = df.iloc[3:6].mean()
mean_4_layers = df.iloc[6:9].mean()
mean_5_layers = df.iloc[9:11].mean()
df_mean.loc['mean_3_layers'] = mean_3_layers
df_mean.loc['mean_4_layers'] = mean_4_layers
df_mean.loc['mean_5_layers'] = mean_5_layers
df_mean.transpose()

<ipython-input-7-ac548c2e8ab5>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_mean.loc['mean_3_layers'] = mean_3_layers
<ipython-input-7-ac548c2e8ab5>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_mean.loc['mean_4_layers'] = mean_4_layers
<ipython-input-7-ac548c2e8ab5>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_mean.loc['mean_5_layers'] = mean_5_layers


test_3_1  test_3_2  test_3_3  test_3_4_layers_64_128_256  \
iou_invalid        0.898628  0.899704  0.900827                    0.927203   
iou_valid          0.927567  0.928168  0.929737                    0.947970   
iou_land           0.998555  0.998078  0.998839                    0.999551   
mean_iou           0.941583  0.941984  0.943134                    0.958241   
precision_invalid  0.953436  0.950507  0.962381                    0.973096   
precision_valid    0.957825  0.961136  0.953584                    0.965816   
precision_land     0.999028  0.998433  0.999259                    0.999611   
mean_precision     0.970096  0.970025  0.971741                    0.979508   
recall_invalid     0.939877  0.943925  0.933705                    0.951597   
recall_valid       0.967064  0.964362  0.973807                    0.980881   
recall_land        0.999526  0.999644  0.999580                    0.999940   
mean_recall        0.968822  0.969310  0.969031                    0.977473   
f1_invalid         0.946608  0.947205  0.947826                    0.962227   
f1_valid           0.962423  0.962746  0.963589                    0.973290   
f1_land            0.999277  0.999038  0.999419                    0.999775   
mean_f1            0.969436  0.969663  0.970278                    0.978431   
mean_accuracy      0.975138  0.975267  0.975888                    0.982507   

                   test_4_1  test_4_2  test_4_3  test_5_1  mean_3_layers  \
iou_invalid        0.908179  0.909002  0.910239  0.904243       0.899720   
iou_valid          0.935095  0.935425  0.936735  0.932276       0.928491   
iou_land           0.998831  0.998509  0.998520  0.997891       0.998491   
mean_iou           0.947368  0.947645  0.948498  0.944803       0.942234   
precision_invalid  0.967426  0.961316  0.965619  0.957190       0.955441   
precision_valid    0.955932  0.960634  0.958691  0.959982       0.957515   
precision_land     0.999158  0.998935  0.999065  0.998635       0.998907   
mean_precision     0.974172  0.973628  0.974458  0.971936       0.970621   
recall_invalid     0.936826  0.943514  0.940727  0.942353       0.939169   
recall_valid       0.977221  0.972712  0.976134  0.969972       0.968411   
recall_land        0.999673  0.999573  0.999454  0.999254       0.999583   
mean_recall        0.971240  0.971933  0.972105  0.970526       0.969054   
f1_invalid         0.951880  0.952332  0.953010  0.949714       0.947213   
f1_valid           0.966459  0.966635  0.967334  0.964951       0.962919   
f1_land            0.999415  0.999254  0.999259  0.998944       0.999245   
mean_f1            0.972585  0.972740  0.973201  0.971203       0.969792   
mean_accuracy      0.977758  0.977826  0.978222  0.976531       0.975431   

                   mean_4_layers  mean_5_layers  
iou_invalid             0.914795       0.907241  
iou_valid               0.939497       0.934505  
iou_land                0.998964       0.998206  
mean_iou                0.951085       0.946651  
precision_invalid       0.967280       0.961404  
precision_valid         0.960794       0.959337  
precision_land          0.999234       0.998850  
mean_precision          0.975769       0.973197  
recall_invalid          0.943979       0.941540  
recall_valid            0.976938       0.973053  
recall_land             0.999729       0.999354  
mean_recall             0.973549       0.971316  
f1_invalid              0.955480       0.951362  
f1_valid                0.968795       0.966143  
f1_land                 0.999481       0.999102  
mean_f1                 0.974585       0.972202  
mean_accuracy           0.979364       0.977377

### 5. Conclusion

The previous table shows that the u-nets with a depth of 4 concolutional layers perform best.

It should be noted, however, that this is 2 or 3 trainigs a little too little to take into account statistical variances.

Nevertheless, the mean intersection over union for all 3 runs is worse than that of the best models in the previous experiment. The reason for this could be statistical variance. In addition, transposed convolutional layers were used for the upsampling instead of UpSampling2D layers like in the previous experiments which might have caused the performance decrease.

Therefore 2 additional training runs have been added afterwards. Both configurations use UpSampling2D instead of Conv2DTranspose layers for the upsampling.
Trainig one is named `up_3` as it's U-Net decoder and encoder contain 3 layers, each with 64, 128 and 256 filters. The second training run is named `up_5` as it's U-Net decoder and encoder contain 5 convolutional layers, each with 64, 128, 256, 512 and 1024 filters. In the following table both runs are compared to the best models of the previous table and the best configuration of experiment 7.


In [6]:
metrics = []
all_files = os.listdir(f'../metrics/{experiment}/')
files = []

for i in range(len(all_files) - 1, -1, -1):
    if 'layers' in all_files[i]:
        files.append(all_files[i])

for metric_file in files:
    with open(f'../metrics/{experiment}/{metric_file}', 'rb') as file:
        metric_dict = pickle.load(file).__dict__
        metrics.append(metric_dict)

In [9]:
df = pd.DataFrame(metrics)

df.index = ['test_up_5', 'val_up_5', 'train_up_5', 'test_up_3', 'val_up_3', 'train_up_3']
df.transpose()

test_up_5  val_up_5  train_up_5  test_up_3  val_up_3  \
iou_invalid         0.908735  0.907428    0.983374   0.923502  0.958031   
iou_valid           0.934268  0.933043    0.986902   0.944912  0.965662   
iou_land            0.998995  0.998924    0.999198   0.999532  0.999598   
mean_iou            0.947332  0.946465    0.989825   0.955982  0.974430   
precision_invalid   0.955454  0.961722    0.990473   0.976975  0.985206   
precision_valid     0.963536  0.957958    0.994313   0.960096  0.977314   
precision_land      0.999573  0.999455    0.999638   0.999591  0.999646   
mean_precision      0.972854  0.973045    0.994808   0.978887  0.987389   
recall_invalid      0.948939  0.941430    0.992764   0.944050  0.972014   
recall_valid        0.968511  0.972881    0.992505   0.983538  0.987805   
recall_land         0.999421  0.999468    0.999560   0.999942  0.999952   
mean_recall         0.972291  0.971260    0.994943   0.975843  0.986590   
f1_invalid          0.952185  0.951468    0.991617   0.960230  0.978566   
f1_valid            0.966017  0.965362    0.993408   0.971676  0.982531   
f1_land             0.999497  0.999462    0.999599   0.999766  0.999799   
mean_f1             0.972567  0.972097    0.994875   0.977224  0.986965   
mean_accuracy       0.977675  0.977420    0.995674   0.981628  0.989124   

                   train_up_3  
iou_invalid          0.927203  
iou_valid            0.947970  
iou_land             0.999551  
mean_iou             0.958241  
precision_invalid    0.973096  
precision_valid      0.965816  
precision_land       0.999611  
mean_precision       0.979508  
recall_invalid       0.951597  
recall_valid         0.980881  
recall_land          0.999940  
mean_recall          0.977473  
f1_invalid           0.962227  
f1_valid             0.973290  
f1_land              0.999775  
mean_f1              0.978431  
mean_accuracy        0.982507

The best performing model contains three different Conv2D layers with filter sizes of 64, 128, and 256. It uses UpSampling2D instead of TransposedConv2D layers to upsample.

Mean intersection over union for the best performing model:
0.955982